### PR TITLE
fix: `utils.UpdateObject` duplication of array members

### LINKED ISSUE #545: bug: `utils.UpdateObject` incorrect result with complex input
Hi @ms-henglu 

I have found a strange bug where the provider's `utils.UpdateObject` func unexpectedly modifies the Azure response to produce an incorrect result.

I have validated that this is definitely a provider issue by confirming the resource's state in Azure and the Terraform state file is correct.

In the below repro code, the Azure policy definition has a embedded ARM template, containing two resources at this path: `.properties.policyRule.then.details/deployment.properties.template.resources[]`

What happens is that the `utils.UpdateObject` func duplicates the zero index and adds an additional resource, making the length three.

I have debugged the provider and isolated the issue to the above func. I have created a test case here:

[link to failing unit test](https://github.com/matt-FFFFFF/terraform-provider-azapi/blob/a139a21d1b55df1dcc99420478c48bf8dc0cecc7/utils/json_test.go#L578-L661)

I have proposed a fix for your consideration in #546

Thank you!

## Repro code

```terraform
terraform {
  required_providers {
    azapi = {
      source  = "Azure/azapi"
      version = "1.14.0"
    }
  }
}

data "azapi_client_config" "current" {}

resource "azapi_resource" "policy_definition" {
  type      = "Microsoft.Authorization/policyDefinitions@2023-04-01"
  name      = "Deploy-Diagnostics-PostgreSQL"
  parent_id = "/subscriptions/${data.azapi_client_config.current.subscription_id}"
  body = {
    "properties" = {
      "description" = "Deploys the diagnostic settings for Database for PostgreSQL to stream to a Log Analytics workspace when any Database for PostgreSQL which is missing this diagnostic settings is created or updated. This policy is superseded by built-in initiative https=//www.azadvertizer.net/azpolicyinitiativesadvertizer/0884adba-2312-4468-abeb-5422caed1038.html.",
      "displayName" = "[Deprecated]: Deploy Diagnostic Settings for Database for PostgreSQL to Log Analytics workspace",
      "metadata" = {
        "alzCloudEnvironments" = [
          "AzureCloud",
          "AzureChinaCloud",
          "AzureUSGovernment"
        ],
        "category"   = "Monitoring",
        "deprecated" = true,
        "source"     = "https://github.com/Azure/Enterprise-Scale/",
        "version"    = "2.0.0-deprecated"
      },
      "mode" = "Indexed",
      "parameters" = {
        "effect" = {
          "allowedValues" = [
            "DeployIfNotExists",
            "Disabled"
          ],
          "defaultValue" = "DeployIfNotExists",
          "metadata" = {
            "description" = "Enable or disable the execution of the policy",
            "displayName" = "Effect"
          },
          "type" = "String"
        },
        "logAnalytics" = {
          "metadata" = {
            "description" = "Select Log Analytics workspace from dropdown list. If this workspace is outside of the scope of the assignment you must manually grant 'Log Analytics Contributor' permissions (or similar) to the policy assignment's principal ID.",
            "displayName" = "Log Analytics workspace",
            "strongType"  = "omsWorkspace"
          },
          "type" = "String"
        },
        "logsEnabled" = {
          "allowedValues" = [
            "True",
            "False"
          ],
          "defaultValue" = "True",
          "metadata" = {
            "description" = "Whether to enable logs stream to the Log Analytics workspace - True or False",
            "displayName" = "Enable logs"
          },
          "type" = "String"
        },
        "metricsEnabled" = {
          "allowedValues" = [
            "True",
            "False"
          ],
          "defaultValue" = "True",
          "metadata" = {
            "description" = "Whether to enable metrics stream to the Log Analytics workspace - True or False",
            "displayName" = "Enable metrics"
          },
          "type" = "String"
        },
        "profileName" = {
          "defaultValue" = "setbypolicy",
          "metadata" = {
            "description" = "The diagnostic settings profile name",
            "displayName" = "Profile name"
          },
          "type" = "String"
        }
      },
      "policyRule" = {
        "if" = {
          "anyOf" = [
            {
              "equals" = "Microsoft.DBforPostgreSQL/flexibleServers",
              "field"  = "type"
            },
            {
              "equals" = "Microsoft.DBforPostgreSQL/servers",
              "field"  = "type"
            }
          ]
        },
        "then" = {
          "details" = {
            "deployment" = {
              "properties" = {
                "mode" = "Incremental",
                "parameters" = {
                  "location" = {
                    "value" = "[field('location')]"
                  },
                  "logAnalytics" = {
                    "value" = "[parameters('logAnalytics')]"
                  },
                  "logsEnabled" = {
                    "value" = "[parameters('logsEnabled')]"
                  },
                  "metricsEnabled" = {
                    "value" = "[parameters('metricsEnabled')]"
                  },
                  "profileName" = {
                    "value" = "[parameters('profileName')]"
                  },
                  "resourceName" = {
                    "value" = "[field('name')]"
                  },
                  "resourceType" = {
                    "value" = "[field('type')]"
                  }
                },
                "template" = {
                  "$schema"        = "http=//schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
                  "contentVersion" = "1.0.0.0",
                  "outputs"        = {},
                  "parameters" = {
                    "location" = {
                      "type" = "String"
                    },
                    "logAnalytics" = {
                      "type" = "String"
                    },
                    "logsEnabled" = {
                      "type" = "String"
                    },
                    "metricsEnabled" = {
                      "type" = "String"
                    },
                    "profileName" = {
                      "type" = "String"
                    },
                    "resourceName" = {
                      "type" = "String"
                    },
                    "resourceType" = {
                      "type" = "String"
                    }
                  },
                  "resources" = [
                    {
                      "apiVersion" = "2021-05-01-preview",
                      "condition"  = "[startsWith(parameters('resourceType'),'Microsoft.DBforPostgreSQL/flexibleServers')]",
                      "dependsOn"  = [],
                      "location"   = "[parameters('location')]",
                      "name"       = "[concat(parameters('resourceName'), '/', 'Microsoft.Insights/', parameters('profileName'))]",
                      "properties" = {
                        "logs" = [
                          {
                            "category" = "PostgreSQLLogs",
                            "enabled"  = "[parameters('logsEnabled')]"
                          }
                        ],
                        "metrics" = [
                          {
                            "category" = "AllMetrics",
                            "enabled"  = "[parameters('metricsEnabled')]",
                            "retentionPolicy" = {
                              "days"    = 0,
                              "enabled" = false
                            },
                            "timeGrain" = null
                          }
                        ],
                        "workspaceId" = "[parameters('logAnalytics')]"
                      },
                      "type" = "Microsoft.DBforPostgreSQL/flexibleServers/providers/diagnosticSettings"
                    },
                    {
                      "apiVersion" = "2021-05-01-preview",
                      "condition"  = "[startsWith(parameters('resourceType'),'Microsoft.DBforPostgreSQL/servers')]",
                      "dependsOn"  = [],
                      "location"   = "[parameters('location')]",
                      "name"       = "[concat(parameters('resourceName'), '/', 'Microsoft.Insights/', parameters('profileName'))]",
                      "properties" = {
                        "logs" = [
                          {
                            "category" = "PostgreSQLLogs",
                            "enabled"  = "[parameters('logsEnabled')]"
                          },
                          {
                            "category" = "QueryStoreRuntimeStatistics",
                            "enabled"  = "[parameters('logsEnabled')]"
                          },
                          {
                            "category" = "QueryStoreWaitStatistics",
                            "enabled"  = "[parameters('logsEnabled')]"
                          }
                        ],
                        "metrics" = [
                          {
                            "category" = "AllMetrics",
                            "enabled"  = "[parameters('metricsEnabled')]",
                            "retentionPolicy" = {
                              "days"    = 0,
                              "enabled" = false
                            },
                            "timeGrain" = null
                          }
                        ],
                        "workspaceId" = "[parameters('logAnalytics')]"
                      },
                      "type" = "Microsoft.DBforPostgreSQL/servers/providers/diagnosticSettings"
                    }
                  ],
                  "variables" = {}
                }
              }
            },
            "existenceCondition" = {
              "allOf" = [
                {
                  "equals" = "true",
                  "field"  = "Microsoft.Insights/diagnosticSettings/logs.enabled"
                },
                {
                  "equals" = "true",
                  "field"  = "Microsoft.Insights/diagnosticSettings/metrics.enabled"
                },
                {
                  "equals" = "[parameters('logAnalytics')]",
                  "field"  = "Microsoft.Insights/diagnosticSettings/workspaceId"
                }
              ]
            },
            "name" = "[parameters('profileName')]",
            "roleDefinitionIds" = [
              "/providers/microsoft.authorization/roleDefinitions/749f88d5-cbae-40b8-bcfc-e573ddc772fa",
              "/providers/microsoft.authorization/roleDefinitions/92aaf0da-9dab-42b6-94a3-d43ce8d16293"
            ],
            "type" = "Microsoft.Insights/diagnosticSettings"
          },
          "effect" = "[parameters('effect')]"
        }
      },
      "policyType" = "Custom"
    }
  }
}
```

## Plan output after successful apply

```
data.azapi_client_config.current: Reading...
data.azapi_client_config.current: Read complete after 1s [id=clientConfigs/subscriptionId=d5ffd04f-25c8-4494-a5de-4e1c707bf600;tenantId=c180f90e-b05a-4e58-81a2-3f6f7d9340e1]
azapi_resource.policy_definition: Refreshing state... [id=/subscriptions/d5ffd04f-25c8-4494-a5de-4e1c707bf600/providers/Microsoft.Authorization/policyDefinitions/Deploy-Diagnostics-PostgreSQL]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # azapi_resource.policy_definition will be updated in-place
  ~ resource "azapi_resource" "policy_definition" {
      ~ body                      = {
          ~ properties = {
              ~ policyRule  = {
                  ~ then = {
                      ~ details = {
                          ~ deployment         = {
                              ~ properties = {
                                  ~ template   = {
                                      ~ resources      = [
                                            {
                                                apiVersion = "2021-05-01-preview"
                                                condition  = "[startsWith(parameters('resourceType'),'Microsoft.DBforPostgreSQL/flexibleServers')]"
                                                dependsOn  = []
                                                location   = "[parameters('location')]"
                                                name       = "[concat(parameters('resourceName'), '/', 'Microsoft.Insights/', parameters('profileName'))]"
                                                properties = {
                                                    logs        = [
                                                        {
                                                            category = "PostgreSQLLogs"
                                                            enabled  = "[parameters('logsEnabled')]"
                                                        },
                                                    ]
                                                    metrics     = [
                                                        {
                                                            category        = "AllMetrics"
                                                            enabled         = "[parameters('metricsEnabled')]"
                                                            retentionPolicy = {
                                                                days    = 0
                                                                enabled = false
                                                            }
                                                            timeGrain       = null
                                                        },
                                                    ]
                                                    workspaceId = "[parameters('logAnalytics')]"
                                                }
                                                type       = "Microsoft.DBforPostgreSQL/flexibleServers/providers/diagnosticSettings"
                                            },
                                          - {
                                              - apiVersion = "2021-05-01-preview"
                                              - condition  = "[startsWith(parameters('resourceType'),'Microsoft.DBforPostgreSQL/flexibleServers')]"
                                              - dependsOn  = []
                                              - location   = "[parameters('location')]"
                                              - name       = "[concat(parameters('resourceName'), '/', 'Microsoft.Insights/', parameters('profileName'))]"
                                              - properties = {
                                                  - logs        = [
                                                      - {
                                                          - category = "PostgreSQLLogs"
                                                          - enabled  = "[parameters('logsEnabled')]"
                                                        },
                                                    ]
                                                  - metrics     = [
                                                      - {
                                                          - category        = "AllMetrics"
                                                          - enabled         = "[parameters('metricsEnabled')]"
                                                          - retentionPolicy = {
                                                              - days    = 0
                                                              - enabled = false
                                                            }
                                                          - timeGrain       = null
                                                        },
                                                    ]
                                                  - workspaceId = "[parameters('logAnalytics')]"
                                                }
                                              - type       = "Microsoft.DBforPostgreSQL/flexibleServers/providers/diagnosticSettings"
                                            },
                                            {
                                                apiVersion = "2021-05-01-preview"
                                                condition  = "[startsWith(parameters('resourceType'),'Microsoft.DBforPostgreSQL/servers')]"
                                                dependsOn  = []
                                                location   = "[parameters('location')]"
                                                name       = "[concat(parameters('resourceName'), '/', 'Microsoft.Insights/', parameters('profileName'))]"
                                                properties = {
                                                    logs        = [
                                                        {
                                                            category = "PostgreSQLLogs"
                                                            enabled  = "[parameters('logsEnabled')]"
                                                        },
                                                        {
                                                            category = "QueryStoreRuntimeStatistics"
                                                            enabled  = "[parameters('logsEnabled')]"
                                                        },
                                                        {
                                                            category = "QueryStoreWaitStatistics"
                                                            enabled  = "[parameters('logsEnabled')]"
                                                        },
                                                    ]
                                                    metrics     = [
                                                        {
                                                            category        = "AllMetrics"
                                                            enabled         = "[parameters('metricsEnabled')]"
                                                            retentionPolicy = {
                                                                days    = 0
                                                                enabled = false
                                                            }
                                                            timeGrain       = null
                                                        },
                                                    ]
                                                    workspaceId = "[parameters('logAnalytics')]"
                                                }
                                                type       = "Microsoft.DBforPostgreSQL/servers/providers/diagnosticSettings"
                                            },
                                        ]
                                        # (5 unchanged attributes hidden)
                                    }
                                    # (2 unchanged attributes hidden)
                                }
                            }
                            name               = "[parameters('profileName')]"
                            # (3 unchanged attributes hidden)
                        }
                        # (1 unchanged attribute hidden)
                    }
                    # (1 unchanged attribute hidden)
                }
                # (6 unchanged attributes hidden)
            }
        }
        id                        = "/subscriptions/d5ffd04f-25c8-4494-a5de-4e1c707bf600/providers/Microsoft.Authorization/policyDefinitions/Deploy-Diagnostics-PostgreSQL"
        name                      = "Deploy-Diagnostics-PostgreSQL"
      ~ output                    = {} -> (known after apply)
        # (6 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

### PR DESCRIPTION
**Pull request by matt-FFFFFF**
fixes #545

`utils.UpdateObject` would duplicate array members under certain conditions.

Adds `reflect.DeepEqual()` check to start of `utils.UpdateObject` to return early if old is the same as new. This could be masking the original bug but it's still probably a worthy thing to do.

This PR adds a test, which used to fail but now passes.

Pinging @ms-henglu

### CODE REVIEW

**Comment by mcleanbc** [comment]
I also ran into the linked issue, using the Microsoft.CognitiveServices/accounts/raiPolicies API, which returns multiple entries with the same value for `name`, e.g.

```
               "contentFilters": [
                    {
                        "name": "Hate",
                        "allowedContentLevel": "Medium",
                        "blocking": true,
                        "enabled": true,
                        "source": "Prompt"
                    },
                    {
                        "name": "Hate",
                        "allowedContentLevel": "Medium",
                        "blocking": true,
                        "enabled": true,
                        "source": "Completion"
                    }, 
```

If the API returns the `contentFilters` array in guaranteed order, the return early solution would work: the array would be deeply equal. But if the array order is not guaranteed (for this API or others with similar treatment of `name`), the issue would reoccur.

Rather than (or in addition to) a return early, I'd recommend replacing the call to `areSameArrayItems` on [ln 105](https://github.com/matt-FFFFFF/terraform-provider-azapi/blob/66650c8495d270697379ef3abb4f02d4a15b43f7/utils/json.go#L105) with `reflect.DeepEqual` which would be more reliable than checking the `name` field, which is the current behavior. https://github.com/matt-FFFFFF/terraform-provider-azapi/blob/66650c8495d270697379ef3abb4f02d4a15b43f7/utils/json.go#L133

**Comment by matt-FFFFFF** [comment]
Thanks for this @mcleanbc 

I'll update my PR with your suggestions and re-test.

**Comment by ms-henglu** [comment]
Thanks @matt-FFFFFF for the PR! And thanks @mcleanbc for the great suggestions!

**Comment by mcleanbc** [comment]
Looks good to me. Thanks @matt-FFFFFF for including those updates!

**Comment by ms-henglu** [comment]
Hello @matt-FFFFFF , thanks for working on this.

But the test case should not be modified. Let me add more contexts here.

The logic uses `name` field as identifier is used to solve the below issue.

User uses azapi_resource to manage a aks cluster, which must have a system node pool defined. And he also uses azapi_resource to add a user node pool. 

```hcl
resource "azapi_resource" "cluster" {
  type      = "Microsoft.ContainerService/managedClusters@2024-04-02-preview"
  parent_id = azapi_resource.resourceGroup.id
  name      = "example"
  location  = azapi_resource.resourceGroup.location
  identity {
    type         = "SystemAssigned"
    identity_ids = []
  }
  body = {
    properties = {
      agentPoolProfiles = [
        {
          count  = 1
          mode   = "System"
          name   = "default"
          vmSize = "Standard_DS2_v2"
        },
      ]
      dnsPrefix = "example"
    }
  }
}

resource "azapi_resource" "pool" {
  type      = "Microsoft.ContainerService/managedClusters/agentPools@2024-04-02-preview"
  parent_id = azapi_resource.cluster.id
  name      = "internal"
  body = {
    properties = {
      count  = 1
      mode   = "User"
      vmSize = "Standard_DS2_v2"
    }
  }
}
```

In the next plan, the `azapi_resource.cluster` will say there's diff in the `body.properties.agentPoolProfiles`, because there're actually two elements in the remote, but only one is defined. 

```bash
 ~ agentPoolProfiles = [
                  ~ {
                      - currentOrchestratorVersion = "1.28.9"
                      - enableFIPS                 = false
                      - kubeletDiskType            = "OS"
                      - maxPods                    = 110
                        name                       = "default"
                      - nodeImageVersion           = "AKSUbuntu-2204gen2containerd-202406.19.0"
                      - orchestratorVersion        = "1.28"
                      - osDiskSizeGB               = 128
                      - osDiskType                 = "Managed"
                      - osSKU                      = "Ubuntu"
                      - osType                     = "Linux"
                      - powerState                 = {
                          - code = "Running"
                        }
                      - provisioningState          = "Succeeded"
                      - type                       = "VirtualMachineScaleSets"
                      - upgradeSettings            = {
                          - maxSurge = "10%"
                        }
                        # (3 unchanged attributes hidden)
                    },
                  - {
                      - count                      = 1
                      - currentOrchestratorVersion = "1.28.9"
                      - enableAutoScaling          = false
                      - enableFIPS                 = false
                      - kubeletDiskType            = "OS"
                      - maxPods                    = 110
                      - mode                       = "User"
                      - name                       = "internal"
                      - nodeImageVersion           = "AKSUbuntu-2204gen2containerd-202406.19.0"
                      - orchestratorVersion        = "1.28"
                      - osDiskSizeGB               = 128
                      - osDiskType                 = "Managed"
                      - osSKU                      = "Ubuntu"
                      - osType                     = "Linux"
                      - powerState                 = {
                          - code = "Running"
                        }
                      - provisioningState          = "Succeeded"
                      - scaleDownMode              = "Delete"
                      - type                       = "VirtualMachineScaleSets"
                      - vmSize                     = "Standard_DS2_v2"
                    },
                ]
```

The goal is to make it easier for users to use `lifecycle.ignore_changes` to suppress the diff. And we don't want users to use `lifecycle.ignore_changes=[body.properties.agentPoolProfiles]` to suppress the whole array which makes it impossible to update the system node pool.

So I use the `name` field as an identifier, to map the system node pool and make it use the same schema as the config. Then the plan-diff will only show there's only one element added, and the system node pool has no diff. Then user can use `lifecycle.ignore_changes=[body.properties.agentPoolProfiles[1]]` to suppress the diff.

```bash
agentPoolProfiles = [
                    {
                        count  = 1
                        mode   = "System"
                        name   = "default"
                        vmSize = "Standard_DS2_v2"
                    },
                  - {
                      - count                      = 1
                      - currentOrchestratorVersion = "1.28.9"
                      - enableAutoScaling          = false
                      - enableFIPS                 = false
                      - kubeletDiskType            = "OS"
                      - maxPods                    = 110
                      - mode                       = "User"
                      - name                       = "internal"
                      - nodeImageVersion           = "AKSUbuntu-2204gen2containerd-202406.19.0"
                      - orchestratorVersion        = "1.28"
                      - osDiskSizeGB               = 128
                      - osDiskType                 = "Managed"
                      - osSKU                      = "Ubuntu"
                      - osType                     = "Linux"
                      - powerState                 = {
                          - code = "Running"
                        }
                      - provisioningState          = "Succeeded"
                      - scaleDownMode              = "Delete"
                      - type                       = "VirtualMachineScaleSets"
                      - vmSize                     = "Standard_DS2_v2"
                    },
                ]
```

File: utils/json.go [hunk 1]
```diff
@@ -102,7 +106,7 @@ func UpdateObject(old interface{}, new interface{}, option UpdateJsonOption) int
 
 			for _, oldItem := range oldValue {
```
Comment (ms-henglu) [inline utils/json.go:107]: How about the below codes:
It tries to find the deeply equaled element first, if not found, use the item with same identifier.

```go
for _, oldItem := range oldValue {
	found := false
	for index, newItem := range newArr {
		if reflect.DeepEqual(oldItem, newItem) && !used[index] {
			res = append(res, UpdateObject(oldItem, newItem, option))
			used[index] = true
			found = true
			break
		}
	}
	if found {
		continue
	}
	for index, newItem := range newArr {
		if areSameArrayItems(oldItem, newItem) && !used[index] {
			res = append(res, UpdateObject(oldItem, newItem, option))
			used[index] = true
			break
		}
	}
}
```

**Comment by matt-FFFFFF** [comment]
Thanks @ms-henglu I have updated the code with your suggestion.

I have reverted the changes to the test case and also added new cases to cover the scenarios raised by myself and @mcleanbc.

All tests now pass.

**Review by ms-henglu — APPROVED** [review]
Thanks for the PR! LGTM

**Comment by matt-FFFFFF** [comment]
Thanks for approving! please can you make a `v1.14.1` patch release if `v1.15.0` is not ready yet?

**Pull request merged**